### PR TITLE
[Feature] Add xquery recursive collection creation with just one path expression

### LINKED ISSUE #4238: [feature] add `mkdir -p` like function to xmldb XQuery module
I've seen one-off implementations - and built one of my own - to solve a common problem:

> Create an collection  where missing parent collections will be __created__.

On the linux command line this is possible by setting the `-p` flag on `mkdir`.

```
mkdir -p `/this/is/a/deeply/nested/directory`
```

nodeJS for example mimics this behaviour by setting the `recursive` option on the builtin `fs.mkdir` method.

There is no such facility present in eXistdb to my knowledge. That is why I propose to add that by either
1. a new function `xmldb:create-collection-recursive` accepting one parameter:
  - collectionUri as xs:string

2. or add a signature to `xmldb:create-collection` accepting two parameters:
  - collectionUri as xs:string
  - options as xs:map
 
3. create a new builtin XQuery module that will handle this and all other collection operations. This module would then deprecate
  currently available functions `xmldb:create-collection`, `xmldb:collection-available` and the like. While this is definitely a controversial proposal it also has the potential to rethink our current approach while maintaining backwards compatibility. 

The fact that xmldb:create-collection already has a two parameter signature does rule out option 2 unless we choose to also
introduce a breaking change.  The current `xmldb:create-collection#2` would have to be changed to 
`xmldb:create-collection($collectionUri as xs:string) as xs:string` removing the parent collection parameter that is currently expected.

I personally favor having the collection URIs to be in one parameter but certainly am open to discuss the motivation behind the current implementation.

### PR DESCRIPTION
**Pull request by enima007**
### Description:

Added an xmldb:create-collection#1 function to XMLDBModule.java and updated XMLDBCreateCollection.java implementation 
to allow the creation of a new collection just by specifying its path.

### Reference:

Closes https://github.com/eXist-db/exist/issues/4238.

### Type of tests:

Added a small assertEqual test in exist-core/src/test/xquery/xmldb/collection-create-tests.xql to check if 
the collection is created successfully.

### CODE REVIEW

File: exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java [hunk 1]
```diff
@@ -25,64 +25,161 @@
...
+				final org.exist.collections.Collection internalCol = ((NodeProxy) node).getOwnerDocument().getCollection();
+				if (logger.isDebugEnabled()) {
+					logger.debug("Found node");
+				}
+				try {
+					//TODO: use xmldbURI
+					collection = getLocalCollection(this, context, internalCol.getURI().toString());
+					return new StringValue(this, collection.getName());
+				} catch (final XMLDBException e) {
+					if (logger.isDebugEnabled()) {
```
Comment (adamretter) [inline exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java:133]: This should be trace level instead

File: exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java [hunk 2]
```diff
@@ -25,64 +25,161 @@
...
+		final Item item = args[paramNumber].itemAt(0);
+		if (Type.subTypeOf(item.getType(), Type.NODE)) {
+			final NodeValue node = (NodeValue) item;
+			if (logger.isDebugEnabled()) {
+				logger.debug("Found node");
+			}
+			if (node.getImplementationType() == NodeValue.PERSISTENT_NODE) {
+				final org.exist.collections.Collection internalCol = ((NodeProxy) node).getOwnerDocument().getCollection();
+				if (logger.isDebugEnabled()) {
+					logger.debug("Found node");
```
Comment (adamretter) [inline exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java:126]: I don't thin we need log this.

File: exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java [hunk 3]
```diff
@@ -25,64 +25,161 @@
...
+			return super.eval(args, contextSequence);
+		}
+
+		final boolean collectionNeedsClose = false;
+		Collection collection = null;
+		final Item item = args[paramNumber].itemAt(0);
+		if (Type.subTypeOf(item.getType(), Type.NODE)) {
+			final NodeValue node = (NodeValue) item;
+			if (logger.isDebugEnabled()) {
+				logger.debug("Found node");
```
Comment (adamretter) [inline exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java:121]: I don't think we need log this.

File: exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java [hunk 4]
```diff
@@ -25,64 +25,161 @@
...
+			} catch (final XMLDBException xe) {
+				if (logger.isDebugEnabled()) {
+					logger.debug("Couldn't find parent collection, creating a new one.");
+				}
+			}
+		} else {
+			try {
+				return new StringValue(this, collection.getName());
+			} catch (XMLDBException e) {
+                throw new RuntimeException(e);
```
Comment (adamretter) [inline exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java:156]: Avoid runtime exception - either throw `XPathException` or `XMLDBException`.

File: exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java [hunk 5]
```diff
@@ -25,64 +25,161 @@
...
+				return new StringValue(this, collection.getName());
+			} catch (final XMLDBException xe) {
+				if (logger.isDebugEnabled()) {
+					logger.debug("Couldn't find parent collection, creating a new one.");
+				}
+			}
+		} else {
+			try {
+				return new StringValue(this, collection.getName());
+			} catch (XMLDBException e) {
```
Comment (adamretter) [inline exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java:155]: exception should be marked `final`

File: exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java [hunk 6]
```diff
@@ -25,64 +25,161 @@
...
+
+		Sequence s = Sequence.EMPTY_SEQUENCE;
+		try {
+			collection = getRootCollection(context, collectionURI);
+			s = evalWithCollection(collection, args, contextSequence);
+		} finally {
+			if (collectionNeedsClose && collection != null) {
+				try {
+					collection.close();
+				} catch (final Exception e) {
```
Comment (adamretter) [inline exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java:168]: Avoid catch-all `Exception`. Instead specify explicitly multiple exceptions separated by `|`

File: exist-core/src/test/xquery/xmldb/collection-create-tests.xql [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+ :)
+xquery version "3.0";
+
+module namespace t="http://exist-db.org/testsuite/collection-create";
+
+import module namespace test="http://exist-db.org/xquery/xqsuite" at "resource:org/exist/xquery/lib/xqsuite/xqsuite.xql";
+
+declare variable $t:parent-collection-name := "parent-collection";
+declare variable $t:parent-collection := "/db/" || $t:parent-collection-name;
+declare variable $t:path-collection := "path/to/new-collection";
```
Comment (adamretter) [inline exist-core/src/test/xquery/xmldb/collection-create-tests.xql:30]: Collection paths should always start with a `/`, and are always started with `/db`...

File: exist-core/src/test/xquery/xmldb/collection-create-tests.xql [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    %test:tearDown
+function t:cleanup() {
+    xmldb:remove($t:parent-collection)
+};
+
+declare
+    %test:assertEquals("/db/path/to/new-collection")
+function t:fnDocAvailableOnHiddenResource() {
+    let $collection := xmldb:create-collection($t:path-collection)
+    return (
```
Comment (adamretter) [inline exist-core/src/test/xquery/xmldb/collection-create-tests.xql:48]: No sequence constructor needed in the `return` expression.

**Review by adamretter — CHANGES_REQUESTED** [review]
Excellent first contribution. Just a few things that can be tidied up.

File: exist-core/src/test/xquery/xmldb/collection-create-tests.xql (hunk 7 above)
Comment (adamretter) [reply exist-core/src/test/xquery/xmldb/collection-create-tests.xql:30]: It is worth checking that your test and code work with `/db/path/to/new-collection`

File: exist-core/src/test/xquery/xmldb/collection-create-tests.xql [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    %test:tearDown
+function t:cleanup() {
+    xmldb:remove($t:parent-collection),
+    xmldb:remove($t:path-collection-from-root)
+};
+
+declare
+    %test:assertEquals("/db/parent-collection/path/to/new-collection")
+function t:fnCreateNewRecursiveCollection() {
+    let $collection := xmldb:create-collection($t:path-collection)
```
Comment (adamretter) [inline exist-core/src/test/xquery/xmldb/collection-create-tests.xql:49]: Do I understand correctly that you are currently testing that:
```
xmldb:create-collection("/parent-collection/path/to/new-collection")
```
creates the collection `/db/parent-collection/path/to/new-collection`?

If so, that should not work! The `xmldb:create-collection#1` function should always take an absolute path, i.e. starting with `/db`. This test should be modified to check for an error result, and the implementation updated accordingly.

**Review by adamretter — CHANGES_REQUESTED** [review]
Paths to `xmldb:create-collection#1` must be absolute.

File: exist-core/src/test/xquery/xmldb/collection-create-tests.xql [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+declare
+    %test:tearDown
+function t:cleanup() {
+    xmldb:remove($t:parent-collection)
+};
+
+declare
+    %test:assertEquals("/db/path/to/new-collection")
+function t:fnDocAvailableOnHiddenResource() {
```
Comment (adamretter) [inline exist-core/src/test/xquery/xmldb/collection-create-tests.xql]: XQuery function names should be "kebab-case"

File: exist-core/src/test/xquery/xmldb/collection-create-tests.xql [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+declare
+    %test:tearDown
+function t:cleanup() {
+    xmldb:remove($t:parent-collection),
+    xmldb:remove($t:path-collection-from-root)
+};
+
+declare
+    %test:assertEquals("/db/parent-collection/path/to/new-collection")
+function t:fnCreateNewRecursiveCollection() {
```
Comment (adamretter) [inline exist-core/src/test/xquery/xmldb/collection-create-tests.xql:48]: XQuery function names should be "kebab-case"

File: exist-core/src/test/xquery/xmldb/collection-create-tests.xql [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    let $collection := xmldb:create-collection($t:path-collection)
+    return $collection
+};
+
+declare
+    %test:assertEquals("/db/path/to/new-collection-from-root")
+function t:fnCreateNewRecursiveCollectionFromRoot() {
```
Comment (adamretter) [inline exist-core/src/test/xquery/xmldb/collection-create-tests.xql:55]: XQuery function names should be "kebab-case"

File: exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java [hunk 13]
```diff
@@ -24,65 +24,182 @@
...
+			return super.eval(args, contextSequence);
+		}
+
+		final Item item = args[paramNumber].itemAt(0);
+		if (Type.subTypeOf(item.getType(), Type.NODE)) {
+			final NodeValue node = (NodeValue) item;
+			if (node.getImplementationType() == NodeValue.PERSISTENT_NODE) {
+				final org.exist.collections.Collection internalCol = ((NodeProxy) node).getOwnerDocument().getCollection();
+				//TODO: use xmldbURI
+				try (Collection collection = getLocalCollection(this, context, internalCol.getURI().toString())) {
```
Comment (adamretter) [inline exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java:140]: Please change:
```java
try (Collection collection
```

to:
```
```java
try (final Collection collection
```

File: exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java [hunk 14]
```diff
@@ -24,65 +24,182 @@
...
+					return new StringValue(this, collection.getName());
+				} catch (final XMLDBException e) {
+					if (logger.isTraceEnabled()) {
+						logger.debug("Couldn't find parent collection, creating a new one.");
+					}
+
+					final String collectionURI = args[paramNumber].getStringValue();
+					if (collectionURI != null) {
+						try (Collection collection = getCollection(this, context, collectionURI, Optional.empty(), Optional.empty())) {
```
Comment (adamretter) [inline exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java:149]: Please change:

try (Collection collection
to:

```java
try (final Collection collection

File: exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java [hunk 15]
```diff
@@ -24,65 +24,182 @@
...
+						}
+					}
+				}
+			} else {
+				return Sequence.EMPTY_SEQUENCE;
+			}
+		}
+
+		Sequence s = Sequence.EMPTY_SEQUENCE;
+		try (Collection rootCollection = getRootCollection(context)) {
```
Comment (adamretter) [inline exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java:164]: Please change:

try (Collection collection
to:

```java
try (final Collection collection

File: exist-core/src/test/xquery/xmldb/collection-create-tests.xql (hunk 9 above)
Comment (line-o) [reply exist-core/src/test/xquery/xmldb/collection-create-tests.xql:50]: looks good to me

File: exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java (hunk 6 above)
Comment (dizzzz) [reply exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java]: code has gone?

File: exist-core/src/test/xquery/xmldb/collection-create-tests.xql (hunk 7 above)
Comment (dizzzz) [reply exist-core/src/test/xquery/xmldb/collection-create-tests.xql]: code looks OK now?

**Review by dizzzz — APPROVED** [review]
LGTM

File: exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java [hunk 16]
```diff
@@ -24,65 +24,182 @@
...
+	public Sequence evalWithCollection(Collection collection, Sequence[] args, Sequence contextSequence)  throws XPathException {
+		final String collectionName;
+		if (args.length > 1) {
+			collectionName = args[1].getStringValue();
+		} else {
+			collectionName = cleanCollectionUri(args[0].getStringValue());
+		}
+
+		try {
+			final Collection newCollection = createCollectionPath(collection, collectionName);
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/xquery/functions/xmldb/XMLDBCreateCollection.java:100]: shoudl this go in to ```try(final Collection newCollection..)``` ?

**Comment by line-o** [comment]
I made an attempt to resurrect this PR with a feature that we would want to have. Unfortunately, there are several todos still in the code and I also wasn't able to fully understand the approach that was chosen.

tl;dr This needs more work before it can be merged.

**Comment by duncdrum** [comment]
see #6666 this is implemented for the #2 version of the function no?

**Comment by line-o** [comment]
Thanks @enima007 for attempting to implement the feature that I was asking for. Please understand that we will close this PR as this is not being worked on and the feature is also incomplete.
